conda_forge_tick/update_prs.py

```diff
@@ -0,0 +1,179 @@
+                closedAt
+                createdAt
+                number
+                title
+                databaseId
+                baseRepository {
+                  name
+                }
+              }
+              pageInfo {
+                hasNextPage
+                endCursor
+              }
+            }
+          }
+        }
+    """)  # noqa
+    headers = {"Authorization": f"token {github_token}"}
+    # Try several times because this times out
+    for i in range(10):
+        logger.info("graphQL request try %d", i + 1)
+        resp = requests.post(
+            'https://api.github.com/graphql',
+            json={'query': query},
+            headers=headers,
+        )
+        if resp.status_code == 200 and 'data' in resp.json():
+            data = resp.json()
+            pr_ids = []
+            for node in data['data']['user']['pullRequests']['nodes']:
+                pr_ids.append(node['databaseId'])
+            return pr_ids
+        time.sleep(10)
+    return []
+
+
+PR_JSON_REGEX = re.compile(r"^pr_json/([0-9]*).json$")
+
+
+def _update_pr(update_function, dry_run, gx):
+    failed_refresh = 0
+    succeeded_refresh = 0
+    gh = "" if dry_run else github_client()
+    futures = {}
+    node_ids = list(gx.nodes)
+    # this makes sure that github rate limits are dispersed
+    random.shuffle(node_ids)
+
+    pr_info_ordered = OrderedDict()
+    if not dry_run:
+        last_prs = _get_last_updated_prs()
+    else:
+        last_prs = []
+    # Setting them here first gives them the highest priority in the OrderedDict
+    for pr_id in last_prs:
+        pr_info_ordered[pr_id] = None
+
+    with executor("thread", NUM_GITHUB_THREADS) as pool:
+        for node_id in tqdm.tqdm(node_ids, desc='ordering PRs', leave=False):
+            node = gx.nodes[node_id]["payload"]
+            prs = node.get("PRed", [])
+            for i, migration in enumerate(prs):
+                pr_json = migration.get("PR", None)
+                # allow for false
+                if pr_json:
+                    if '__lazy_json__' in pr_json:
+                        m = PR_JSON_REGEX.match(pr_json['__lazy_json__'])
+                        if m:
+                            pr_id = int(m.group(1))
+                        else:
+                            pr_id = object()
+                    else:
+                        pr_id = object()
+                    pr_info_ordered[pr_id] = (pr_json, node_id, i)
+
+        for pr_id, v in pr_info_ordered.items():
+            if v:
+                (pr_json, node_id, i) = v
+                future = pool.submit(update_function, ghctx, pr_json, gh, dry_run)
+                futures[future] = (node_id, i, pr_json)
+
+        for f in as_completed(futures):
+            name, i, pr_json = futures[f]
+            try:
+                res = f.result()
+                if res:
+                    succeeded_refresh += 1
+                    pr_json.update(**res)
```

Comment:
Just to make sure, this call issues a write to disk via the lazy Json?

Reply:
Yes, `__setitem__` has a dump at the end of it.